File: typescript/src/advertisements/user/domain/SupervisorUser.ts
import {UserBase} from "./UserBase";
import {UserId} from "../../shared/domain/value-object/UserId";
import {Email} from "../../shared/domain/value-object/Email";
import {Role} from "./value-object/Role";
import {UserStatus} from "./value-object/UserStatus";
import {InvalidUserException} from "./exceptions/InvalidUserException";

export class SupervisorUser extends UserBase {
    protected constructor(id: UserId, email: Email, role: Role, status: UserStatus) {
        if (role !== Role.SUPERVISOR) {
            throw InvalidUserException.build('Invalid role for supervisor user');
        }
        super(id, email, role, status);
    }

    public static fromDatabase(id: UserId, email: Email, role: Role, status: UserStatus): SupervisorUser {
        return new SupervisorUser(id, email, role, status);
    }
}

File: kotlin/src/main/kotlin/advertisements/user/domain/SupervisorUser.kt
package advertisements.user.domain

import advertisements.shared.value_object.Email
import advertisements.shared.value_object.UserId
import advertisements.user.domain.value_object.Role
import advertisements.user.domain.value_object.UserStatus
import advertisements.user.domain.exceptions.InvalidUserException

class SupervisorUser private constructor(
    id: UserId,
    email: Email,
    role: Role,
    status: UserStatus
) : UserBase(id, email, role, status) {

    init {
        if (role != Role.SUPERVISOR) {
            throw InvalidUserException.build("Invalid role for supervisor user")
        }
    }

    companion object {
        @Throws(InvalidUserException::class)
        fun fromDatabase(id: UserId, email: Email, role: Role, status: UserStatus): SupervisorUser {
            return SupervisorUser(id, email, role, status)
        }
    }
}